### PR TITLE
Merger the .h and .cc of all plugins in RecoParlicleFlow/PFProducer

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

Before we continue with all the plugins .h and .cc files getting merged one-by-one with every PR that touches any specific plugin, I simply propose to do that for all of them in one go.

#### PR validation:

CMSSW compiles.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
(I still miss the real advantage of doing so... I also don't see counter-indications, though)

**Comment by guitargeek** [comment]
Hi @perrotta, the advantages are for example:
* you need to open and edit less files when changing a plug-in
* includes are less of a mess because they are at one place instead of two
* the missing header file indicates no other plugin is inheriting from this

Disadvantages: none I think.

The new code guidelines also recommend having plugins in one .cc file, which makes perfect sense to me.

File: RecoParticleFlow/PFProducer/plugins/PFCandidateChecker.cc [hunk 1]
```diff
@@ -1,13 +1,86 @@
...
-
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/Utilities/interface/Exception.h"
 #include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDAnalyzer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFCandidateChecker.cc]: Remove duplicate inclusion (and perhaps also arrange by putting DataFormat and FWCore headers together

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 2]
```diff
@@ -15,41 +13,173 @@
...
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlockElementFwd.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/EgammaReco/interface/BasicClusterFwd.h"
+#include "DataFormats/EgammaReco/interface/PreshowerClusterFwd.h"
+#include "DataFormats/EgammaReco/interface/SuperClusterFwd.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonCore.h"
+#include "DataFormats/EgammaCandidates/interface/Photon.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonCoreFwd.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonFwd.h"
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: Duplicate header
Also in this class some rearrangement of the headers which puts together similar ones can be useful

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 3]
```diff
@@ -15,41 +13,173 @@
...
+#include "DataFormats/CaloRecHit/interface/CaloClusterFwd.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: Duplicate

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 4]
```diff
@@ -15,41 +13,173 @@
...
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: Duplicate

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 5]
```diff
@@ -15,41 +13,173 @@
...
+#include "RecoEcal/EgammaCoreTools/interface/PositionCalc.h"
+#include "DataFormats/CaloTowers/interface/CaloTowerCollection.h"
+#include "RecoEcal/EgammaCoreTools/interface/PositionCalc.h"
+#include "RecoEgamma/PhotonIdentification/interface/PhotonIsolationCalculator.h"
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterFunctionFactory.h"
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterFunctionBaseClass.h" 
+#include "CondFormats/EcalObjects/interface/EcalFunctionParameters.h" 
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterTools.h"
+#include "DataFormats/EgammaCandidates/interface/ConversionFwd.h"
+#include "DataFormats/Math/interface/Vector3D.h"
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: This one and the three headers next are duplicate

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 6]
```diff
@@ -15,41 +13,173 @@
...
+  // the PF MVA and regression 
+  std::vector<float> pfPhotonMva_;
+  std::vector<float> energyRegression_;
+  std::vector<float> energyRegressionError_;
+
+  //Vector of vector of Conversions Refs
+  std::vector<reco::ConversionRefVector > pfConv_;
+  std::vector< std::vector<reco::TrackRef> > pfSingleLegConv_;
+  std::vector< std::vector<float> > pfSingleLegConvMva_;
+  //std::vector<reco::TrackRef> * pfSingleLegConv_;
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: Remove if not needed

File: RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc [hunk 7]
```diff
@@ -15,41 +13,173 @@
...
+  void createSuperClusters(const reco::PFCandidateCollection &,
+			  reco::SuperClusterCollection &superClusters) const;
+  
+  void createOneLegConversions(const edm::OrphanHandle<reco::SuperClusterCollection> & superClustersHandle, reco::ConversionCollection &oneLegConversions);
+
+  //create photon cores
+  void createPhotonCores(const edm::OrphanHandle<reco::SuperClusterCollection> & superClustersHandle, const edm::OrphanHandle<reco::ConversionCollection> & oneLegConversionHandle, reco::PhotonCoreCollection &photonCores) ;
+
+  //create photons
+  //void createPhotons(reco::VertexCollection &vertexCollection, const edm::OrphanHandle<reco::PhotonCoreCollection> & superClustersHandle, const CaloTopology* topology, const EcalRecHitCollection * barrelRecHits, const EcalRecHitCollection * endcapRecHits, const edm::Handle<CaloTowerCollection> & hcalTowersHandle, const IsolationValueMaps& isolationValues, reco::PhotonCollection &photons) ;
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/PFPhotonTranslator.cc]: Remove if not needed

**Review by perrotta — COMMENTED** [review]
Please profit of this PR and clean up a bit where needed.
A few suggestions are listed below.

Further, I would run clang-format on the files touched here (in a separate commit, after having committed the fixes suggested here): while looking through them I notice that there is clearly large room for improvement in several parts of these files, but it is not worth doing all that by hand.

**Comment by guitargeek** [comment]
I tired again to use the include-what-you-use tool that @slava77 recommended to me one day to clean up the includes:
```
scram b -k -j 6 compile COMPILER=iwyu
```
Unfortunately it does not work anymore, I get this error:
```
error: unable to handle compilation, expected exactly one compiler job in ''
```
Anyone has seen that before? Am I doing it wrong?

**Comment by perrotta** [comment]
> I tired again to use the include-what-you-use tool that @slava77 recommended to me one day to clean up the includes:
> 
> ```
> scram b -k -j 6 compile COMPILER=iwyu
> ```
> 
> Unfortunately it does not work anymore, I get this error:
> 
> ```
> error: unable to handle compilation, expected exactly one compiler job in ''
> ```
> 
> Anyone has seen that before? Am I doing it wrong?

Uhm, I get the same output and I have no idea... Maybe @smuzaffar can help?

**Comment by davidlange6** [comment]
does it work without the -j 6?

**Comment by smuzaffar** [comment]
@guitargeek , please ignore that error. It still does run right and suggests things like
```
RecoParticleFlow/PFProducer/plugins/linkers/ECALAndHCALCaloJetLinker.cc should add these lines:
#include <memory>
#include <string>
#include "DataFormats/Common/interface/RefTraits.h"
#include "DataFormats/ParticleFlowReco/interface/PFBlockElement.h"
#include "DataFormats/ParticleFlowReco/interface/PFClusterFwd.h"
#include "FWCore/ParameterSet/interface/ParameterSet.h"
#include "FWCore/PluginManager/interface/PluginFactory.h"
#include "FWCore/Utilities/interface/Exception.h"

/build/muz/x/CMSSW_11_0_X_2019-06-11-2300/src/RecoParticleFlow/PFProducer/plugins/linkers/ECALAndHCALCaloJetLinker.cc should remove these lines:

The full include-list for /build/muz/x/CMSSW_11_0_X_2019-06-11-2300/src/RecoParticleFlow/PFProducer/plugins/linkers/ECALAndHCALCaloJetLinker.cc:
#include <memory>
#include <string>
#include "DataFormats/Common/interface/RefTraits.h"
#include "DataFormats/ParticleFlowReco/interface/PFBlockElement.h"
#include "DataFormats/ParticleFlowReco/interface/PFBlockElementCluster.h"
#include "DataFormats/ParticleFlowReco/interface/PFCluster.h"
#include "DataFormats/ParticleFlowReco/interface/PFClusterFwd.h"
#include "FWCore/ParameterSet/interface/ParameterSet.h"
#include "FWCore/PluginManager/interface/PluginFactory.h"
#include "FWCore/Utilities/interface/Exception.h"
#include "RecoParticleFlow/PFClusterTools/interface/LinkByRecHit.h"
#include "RecoParticleFlow/PFProducer/interface/BlockElementLinkerBase.h"
```

**Comment by guitargeek** [comment]
Where do you get this output? I run `scram b -k compile COMPILER=iwyu` without the j 6 now and this is all that I get as output:
```
****WARNING: You have selected iwyu as compiler but there is no iwyu-ccompiler tool setup. Default compiler gcc-ccompiler will be used to comple C files
****WARNING: You have selected iwyu as compiler but there is no iwyu-f77compiler tool setup. Default compiler gcc-f77compiler will be used to comple F77 files
error: unable to handle compilation, expected exactly one compiler job in ''
>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package RecoParticleFlow/PFProducer
>> Entering Package RecoParticleFlow/PFClusterTools
>> Compile sequence completed for CMSSW CMSSW_11_0_X_2019-06-09-0000
```
Also `git status` does not tell me of any new file where these suggestions might have went.

**Comment by guitargeek** [comment]
My bad! Of course you get no output if you already compiled the packages before... I just needed to do a `scram b distclean` and now it's alright. Thanks!

**Comment by guitargeek** [comment]
There are 3 new commits now: one for the includes, one for clang-format and one for the two suggestions Andrea made.

**Comment by perrotta** [comment]
> There are 3 new commits now: one for the includes, one for clang-format and one for the two suggestions Andrea made.

Thank you @guitargeek

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
The HLTFake1 AddOn test failed because of a file access problem in /store, a reason which is totally unrelated to this PR

**Comment by perrotta** [comment]
please test
(retry)

**Comment by perrotta** [comment]
+1
- Technical
- Merging of the files implemented as planned
- Jenkins tests pass and show no differences

**Comment by fabiocos** [comment]
@perrotta @guitargeek I assume that also here cleaning ```cout``` statements was not the main target of the PR...

**Comment by guitargeek** [comment]
No, I thinks that's enough for one PR

**Comment by fabiocos** [comment]
In my check I see:

```
Testing PR #  27159  for branch  master
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerPSEcal.cc  modified in PRs #  ['27049', '27159']
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerPSEcal.h  modified in PRs #  ['27049', '27159']
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackEcal.cc  modified in PRs #  ['27049', '27159']
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackEcal.h  modified in PRs #  ['27049', '27159']
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackHcal.cc  modified in PRs #  ['27049', '27159']
File  RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackHcal.h  modified in PRs #  ['27049', '27159']
```

From a quick look at the changes in both PRs we should have conflicts between this PR and #27049. @guitargeek How would you like to manage this?

**Comment by guitargeek** [comment]
I think https://github.com/cms-sw/cmssw/pull/27049 is on ice until https://github.com/cms-sw/cmssw/pull/27160 will be merged, and then it has to be rebased anyway. So I would suggest to start by merging this PR here and then I wait.

**Comment by fabiocos** [comment]
+1

@guitargeek I move forward with this PR as suggested

**Comment by guitargeek** [comment]
Thanks!

**Pull request merged**